File: src/validate/no_bad_checksums.rs
use sha2::Digest;
use tokio::fs;

use crate::bundle_struct::Bundle;

pub async fn run(base: &str, bundle: &Bundle) -> Result<(), String> {
    let full_path = format!("{base}/{}", bundle.dashboard.path);

    match check_checksum(&full_path, bundle.dashboard.sha256.as_deref()).await {
        Ok(_) => (),
        Err(e) => {
            return Err(format!(
                "Failed dashboard checksum: full_path={full_path} error={e}"
            ))
        }
    }

    for t in &bundle.tables {
        for tt in &t.transforms {
            let full_path = format!("{base}/{}", tt.path);
            match check_checksum(&full_path, tt.sha256.as_deref()).await {
                Ok(_) => (),
                Err(e) => {
                    return Err(format!(
                        "Failed transformation checksum: path={full_path} error={e}"
                    ))
                }
            }
        }
    }

    Ok(())
}

async fn check_checksum(file_path: &str, checksum: Option<&str>) -> Result<(), String> {
    let content = match fs::read_to_string(file_path).await {
        Ok(v) => v,
        Err(e) => {
            return Err(format!(
                "ERROR: {}.{} Failed to read file_path={}: error={}",
                file!(),
                line!(),
                file_path,
                e
            ));
        }
    };

    // If there is a SHA256 then verify it
    if let Some(v) = checksum {
        let sha256 = generate_sha256(&content);
        if sha256 != v {
            return Err(format!(
                "ERROR: {}.{} SHA256 {} does not match for local file {}",
                file!(),
                line!(),
                sha256,
                file_path
            ));
        }
    }
    Ok(())
}

/// Generate a SHA256 hash from a string
fn generate_sha256(input: &str) -> String {
    let mut hasher = sha2::Sha256::new();
    hasher.update(input.as_bytes());
    let result = hasher.finalize();
    format!("{:x}", result)
}

File: bundle-deployer/src/validation/no_bad_checksums.ts
// Validation: Verify SHA256 checksums of files

import type { Bundle } from "../types/bundle.ts";

export async function run(base: string, bundle: Bundle): Promise<void> {
  // Check dashboard checksum
  const dashboardPath = `${base}/${bundle.dashboard.path}`;
  await checkChecksum(dashboardPath, bundle.dashboard.sha256);
  
  // Check transform checksums
  for (const table of bundle.tables) {
    for (const transform of table.transforms) {
      const fullPath = `${base}/${transform.path}`;
      await checkChecksum(fullPath, transform.sha256);
    }
  }
  
  // Check summary table checksums
  if (bundle.summary_tables) {
    for (const summaryTable of bundle.summary_tables) {
      const fullPath = `${base}/${summaryTable.sql.path}`;
      await checkChecksum(fullPath, summaryTable.sql.sha256);
    }
  }
}

async function checkChecksum(filePath: string, checksum?: string): Promise<void> {
  const content = await Deno.readTextFile(filePath);
  
  // If checksum is provided, verify it
  if (checksum) {
    const calculatedHash = await generateSha256(content);
    
    if (calculatedHash !== checksum) {
      throw new Error(
        `SHA256 ${calculatedHash} does not match for local file ${filePath}`
      );
    }
  }
}

async function generateSha256(input: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(input);
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
}